### PR TITLE
Molecular Dynamics example

### PR DESCRIPTION
**Pull request by ktangsali**
# Modulus Pull Request

## Description
Adds a Molecular Dynamics example for Lennard Jones system as defined [here](https://arxiv.org/pdf/2112.03383.pdf).

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: examples/molecular_dynamics/utils.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+            # Load the .npz file
+            with np.load(filepath, "rb") as data:
+                forces = data["forces"].astype(np.float32)
+                all_forces.extend(forces.reshape(1, -1))
+
+    force_mean = np.mean(np.array(all_forces))
+    force_sd = np.std(np.array(all_forces))
+
+    return force_mean, force_sd
```
Comment (dallasfoster) [inline examples/molecular_dynamics/utils.py:41]: So I am not sure about this function and if it has any physical relevance. The mean of all of the forces should be 0 anyways. Is this normalization done in the paper? Would it make more sense to normalize batches of neighborhoods instead of the entire domain?

File: examples/molecular_dynamics/lennard_jones_system.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                    .to(dist.device)
+                )
+
+                out = model(node_fea, edge_fea, g)
+                true_out = torch.tensor(
+                    (forces - force_mean) / force_sd, dtype=torch.float32
+                ).to(dist.device)
+
+                optimizer.zero_grad()
+                loss = F.l1_loss(out, true_out) + 0.001 * torch.mean(out).abs()
```
Comment (dallasfoster) [inline examples/molecular_dynamics/lennard_jones_system.py:179]: Any commentary on this loss function? I think it would be helpful to the user to point out the origin/interpretation of this loss.

File: examples/molecular_dynamics/lennard_jones_system.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                        out_np = out.detach().cpu().numpy()
+                        true_out_np = true_out.detach().cpu().numpy()
+                        forces_pair.append((out_np, true_out_np))
+
+                        # angle computation
+                        dot_product = np.sum(out_np * true_out_np, axis=1)
+                        out_np_mag = np.linalg.norm(out_np, axis=1)
+                        true_out_np_mag = np.linalg.norm(true_out_np, axis=1)
+
+                        cosine = dot_product / (out_np_mag * true_out_np_mag)
```
Comment (dallasfoster) [inline examples/molecular_dynamics/lennard_jones_system.py:244]: It would also be useful to have some commentary on the validation metrics.

File: examples/molecular_dynamics/lennard_jones_system.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                        cosines.append(cosine)
+
+                    plt.clf()
+                    plt.figure(figsize=(5, 5))
+                    for force_system in forces_pair:
+                        pred, true = force_system
+                        pred_mag = pred[:, 0] + pred[:, 1] + pred[:, 2]
+                        true_mag = true[:, 0] + true[:, 1] + true[:, 2]
```
Comment (dallasfoster) [inline examples/molecular_dynamics/lennard_jones_system.py:252]: Again, can we have commentary on this? I am not sure how useful it is to sum the unsigned force components. Wouldn't I want the magnitude instead?

**Review by dallasfoster — COMMENTED** [review]
Only a few minor comments about documentation, otherwise looks good.

File: examples/molecular_dynamics/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Molecular Dynamics using GNNs
+
+This example demonstrates the use of MeshGrapNet from Modulus applied to a Molecular
+Dynamics use case.
+The example solves the Lennard Jones System as described in the [paper here](https://arxiv.org/abs/2112.03383).
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/molecular_dynamics/README.md]: Incorporated the revised description. I agree, the latter sounds better.

File: examples/molecular_dynamics/conf/config.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+model: 
+  input_dim_nodes: 1
+  input_dim_edges: 4
+  output_dim: 3
+  processor_size: 4
+  mlp_activation_fn: "gelu"
+  num_layers_node_processor: 2
+  num_layers_edge_processor: 2
+  num_layers_node_decoder: 2
+  hidden_dim_edge_encoder: 128
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/molecular_dynamics/conf/config.yaml]: Added, can you check if that suffices, or more details are needed?

File: examples/molecular_dynamics/conf/config.yaml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+wb_artifacts: False
+
+max_epochs: 35
+
+lr:
+  start_lr: 0.0003
+  gamma: 0.9999733124642265
+
+distance_threshold: 7.5
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/molecular_dynamics/conf/config.yaml]: Added

File: examples/molecular_dynamics/lennard_jones_system.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    get_rotation_matrix,
+    create_edges,
+    compute_mean_var,
+)
+
+
+def prepare_input(pos, forces, box_size, rotation_matrix=None, add_random_noise=False):
+    """
+    Perform transformations on the input for data augmentation
+    """
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/molecular_dynamics/lennard_jones_system.py:68]: Added

File: examples/molecular_dynamics/lennard_jones_system.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    if rotation_matrix is not None:
+        pos = pos - off
+        pos = np.matmul(pos, rotation_matrix)
+        pos += off
+        forces = np.matmul(forces, rotation_matrix)
+
+    if add_random_noise:
+        pos = pos + np.random.randn(*pos.shape) * 0.005
+    off_2 = np.min(pos, axis=0)
+    pos = pos - off_2
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/molecular_dynamics/lennard_jones_system.py:84]: Great point. added a note about the rotation. Regarding the noise making the points out of the box, I think that's a valid concern, but with the noise magnitude shown above, that should not be the case mostly. When you say map them back, are you suggesting another call of `np.mod()` ?

File: examples/molecular_dynamics/utils.py (hunk 1 above)
Comment (ktangsali) [reply examples/molecular_dynamics/utils.py:41]: Yeah, the mean will be zero always. I think what you are suggesting can potentially make more sense, but then I think the normalization code might be more complex? I just followed the one from the paper currently. I think this normalization is primarily helpful to scale down the forces to easier network prediction, so maybe its fine if it does not have too much of a physical significance? What do you think ?

File: examples/molecular_dynamics/lennard_jones_system.py (hunk 2 above)
Comment (ktangsali) [reply examples/molecular_dynamics/lennard_jones_system.py:212]: Thanks, added!

File: examples/molecular_dynamics/lennard_jones_system.py (hunk 3 above)
Comment (ktangsali) [reply examples/molecular_dynamics/lennard_jones_system.py]: Added

File: examples/molecular_dynamics/lennard_jones_system.py (hunk 4 above)
Comment (ktangsali) [reply examples/molecular_dynamics/lennard_jones_system.py]: Good point. Magnitude makes sense to me too, but the publication plot shows +ve and -ve components indicating that it is the total force. Plotting vector quantities on a 2d plot is not the best way, but I wanted to keep an easy way to reproduce plots

File: examples/molecular_dynamics/lennard_jones_system.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dallasfoster) [reply examples/molecular_dynamics/lennard_jones_system.py:84]: Yes, I am suggesting another call to `np.mod()`.

[the remark this answers is not in the document]
Comment (dallasfoster) [reply examples/molecular_dynamics/lennard_jones_system.py:84]: Something that is besides the point here but is typically computed is what is known as a [`Radial Distribution Function (rdf)`](https://en.wikipedia.org/wiki/Radial_distribution_function), which is essentially a plot of the distribution of interatomic distances. Something like that is helpful to know how close the atoms really are to each other. If atoms under a lennard jones potential are less than about 1 lj distance away from each other then this kind of noise can make the change in force dramatic.

File: examples/molecular_dynamics/utils.py (hunk 1 above)
Comment (dallasfoster) [reply examples/molecular_dynamics/utils.py:41]: It depends, what I am suggesting is really just Batch Normalization (assuming that the model operates on neighborhoods of atoms).

**Review by dallasfoster — APPROVED** [review]

File: examples/molecular_dynamics/lennard_jones/README.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
+# Molecular Dynamics using GNNs
+
+This example demonstrates the use of MeshGrapNet from Modulus applied to a Molecular
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/molecular_dynamics/lennard_jones/README.md]: Thanks! Updated.

File: examples/molecular_dynamics/lennard_jones/README.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+```bash
+python lennard_jones_system.py
+```
+
+Distributed Data Parallel training is enabled for this example. To run the example on
+multiple GPUs, run
+
+```bash
+mpirun -np <num_GPUs> python lennard_jones_system.py
+```
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply examples/molecular_dynamics/lennard_jones/README.md:54]: Thanks, added.
